---
title: "Bayes Factor"
author: "Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Bayes Factor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Goal

Measure the Bayes factor between JC69 and GTR.

Set a random seed:

```{r}
set.seed(314)
```


## Prepare

For those started from a clean slate, run these statements:

```
# Install babette
devtools::install_github("ropensci/babette")

# Install BEAST2 to a default location
beastier::install_beast2()

# Install raket
devtools::install_github("richelbilderbeek/raket")
```

Obtain the crown age of a phylogeny:

```{r}
get_phylogeny_crown_age <- function(phylogeny) {
  n_taxa <- length(phylogeny$tip.label)
  ape::dist.nodes(phylogeny)[n_taxa + 1][1]
}
testit::assert(
  get_phylogeny_crown_age(ape::read.tree(text = "((A:1, B:1):1, C:2);")) == 2
)
testit::assert(
  get_phylogeny_crown_age(ape::read.tree(text = "((A:2, B:2):1, C:3);")) == 3
)
testit::assert(
  get_phylogeny_crown_age(ape::read.tree(text = "((A:2, B:2):2, C:4);")) == 4
)
```
Interpretation of the Bayes factor [2]:

```{r}
interpret_bayes_factor <- function(bayes_factor) {
  if (bayes_factor < 10^0.0) {
    "in favor of other model"
  } else if (bayes_factor < 10^0.5) {
    "barely worth mentioning"
  } else if (bayes_factor < 10^1.0) {
    "substantial"
  } else if (bayes_factor < 10^1.5) {
    "strong"
  } else if (bayes_factor < 10^2.0) {
    "very strong"
  } else {
    "decisive"
  }
}
testit::assert(interpret_bayes_factor(0.5) == "in favor of other model")
testit::assert(interpret_bayes_factor(1.5) == "barely worth mentioning")
testit::assert(interpret_bayes_factor(8.5) == "substantial")
testit::assert(interpret_bayes_factor(12.5) == "strong")
testit::assert(interpret_bayes_factor(85.0) == "very strong")
testit::assert(interpret_bayes_factor(123.0) == "decisive")
```

## Overview

 * Create a data set that *is* JC69
 * Create a Bayesian posterior assuming JC69
 * Create a Bayesian posterior assuming GTR
 * Compare these posteriors using the Bayes factor
 * Interpret the Bayes factor

## Methods

Create an articifical PBD phylogeny with a long time to speciate. That is,
a tree with all speciation event observed early in time (close to the root):
 
```{r fig.width=7}
phylogeny <- ape::read.tree(text = "((((A:12, B:12):1,C:13):1,D:14):1, E:15);")
crown_age <- get_phylogeny_crown_age(phylogeny)
ape::plot.phylo(phylogeny)
```

Create JC69 alignment and save it to a FASTA file:

```{r fig.width=7}
sequence_length <- 1000

alignment_phydat <- phangorn::simSeq(
  phylogeny,
  l = sequence_length,
  rate = 1.0 / crown_age,
  rootseq = rep(c('a', 'c', 'g', 't'), each = sequence_length / 4)
)
testit::assert(class(alignment_phydat) == "phyDat")
alignment_dnabin <- ape::as.DNAbin(alignment_phydat)

fasta_filename <- tempfile(pattern = "bayes_factor_", fileext = ".fasta")
phangorn::write.phyDat(
  alignment_dnabin,
  file = fasta_filename,
  format = "fasta"
)
image(alignment_dnabin)
```

From that alignment, create two posteriors, each assuming their own site model:

```{r}
mcmc <- beautier::create_mcmc(chain_length = 10000)
burn_in_fraction <- 0.2

pos_jc69 <- tracerer::remove_burn_ins(
  babette::bbt_run(
    fasta_filename, 
    site_model = beautier::create_jc69_site_model(),
    mcmc = mcmc
  )$estimates,
  burn_in_fraction = burn_in_fraction
)
knitr::kable(head(pos_jc69))
```

```{r}
pos_gtr <- tracerer::remove_burn_ins(
  babette::bbt_run(
    fasta_filename, 
    site_model = beautier::create_gtr_site_model(),
    mcmc = mcmc
  )$estimates,
  burn_in_fraction = burn_in_fraction
)
knitr::kable(head(pos_gtr))
```

The Bayes factor (usually denoted as K) can be used for model comparison. 
It can be calculated from the marginal likelihoods ratio:

We follow the methods described in the BEAST book [1] 
The Bayes factor equals the likelihood odds iff both models are equally likely.
From an MCMC run, we can simply take the harmonic mean likelihood [1]:

```{r}
# bayes_factor <- rkt_calc_bf(pos_jc69$likelihood, pos_gtr$likelihood)
# print(bayes_factor)
```

Interpreting our results:

```{r}
# print(paste("in favor of model JC69:", interpret_bayes_factor(bayes_factor)))
# print(paste("in favor of model GTR:", interpret_bayes_factor(1.0 / bayes_factor)))
```

## Conclusion

Whatever I (RJCB) have tried, the Bayes factor barely deviated from 1.0.

## References

 * [1] Drummond, Alexei J., and Remco R. Bouckaert. Bayesian evolutionary analysis with BEAST. Cambridge University Press, 2015. Chapter 1.5.6: Bayesian model selection and model fit, page 18
 * [2] H. Jeffreys (1961). The Theory of Probability (3rd ed.). Oxford. p. 432
 * [3] Newton, Michael A., and Adrian E. Raftery. "Approximate Bayesian inference with the weighted likelihood bootstrap." Journal of the Royal Statistical Society. Series B (Methodological) (1994): 3-48.
 * [4] Gelman, Andrew, and Donald B. Rubin. "Avoiding model selection in Bayesian social research." Sociological methodology 25 (1995): 165-173.
 